src/main/java/org/broadinstitute/hellbender/engine/PathURI.java

```diff
@@ -73,6 +76,100 @@ default String getScheme() {
+     * (i.e. "/") or ".", or if the last component does not contain a ".".
+     */
+    default String getExtension() {
+        final String hierarchicalPath = getURI().getPath();
+        final int indexOfLastComponent = hierarchicalPath.lastIndexOf(FileSystems.getDefault().getSeparator());
+        if (indexOfLastComponent != -1 && indexOfLastComponent < hierarchicalPath.length() - 1) {
+            final String lastComponent = hierarchicalPath.substring(indexOfLastComponent + 1);
+            if (lastComponent.length() > 0) {
+                final int indexOfLastDot = lastComponent.lastIndexOf('.');
+                if (indexOfLastDot != -1 && indexOfLastDot < lastComponent.length() - 1) {
+                    // return a string that includes the leading "." to enable easy comparison with the many
+                    // internal file extension constants we have that include the leading "." (i.e., in htsjdk),
+                    // and also for API consistency (since hasExtension() requires the candidate extension to
+                    // include a leading ".", this allows hasExtension(getExtension()) to always work whenever
+                    // getExtension() succeeds
+                    return lastComponent.substring(indexOfLastDot);
+                }
+            }
+        }
+        throw new IllegalArgumentException(String.format("Input path (%s) has no extension", this));
+    }
+
+    /**
+     * Return true if the path component (the hierarchical part of the scheme specific part of the underlying URI)
+     * ends with the provided {@code extension} string.
+     *
+     * @param extension the target extension to test, INCLUDING the leading ".". May not be null.
+     * @return true if the path component of this specifier ends with the extension, otherwise false
+     */
+    default boolean hasExtension(final String extension) {
+        Utils.nonNull(extension, "Target extension must not be null");
+        Utils.validateArg(extension.length() > 1, "Target extension must be length > 1");
+        Utils.validateArg(extension.charAt(0) == '.', "Target extension must include the leading '.'");
+
+        // We don't want to use {@code #getExtension} here, since it won't work correctly if we're comparing an
+        // extension that uses multiple . chars, such as .fasta.gz., and {@code #getExtension} will throw if there
+        // is no extension.
+        return getURI().getPath().toLowerCase().endsWith(extension.toLowerCase());
+    }
+
+    /**
+     * @return the base name (the last component of the hierarchical part of the scheme-specific part of the URI,
+     * after the last "/"), up to but not including the extension (the last ".").
+     * @throws IllegalArgumentException if the last component is empty (ie, the component ends in "/"), or the last
+     * component exists but starts with "."
+     */
+    default String getBaseName() {
```

Comment:
We might want to prefer returning empty string when the base name/ extension are empty.   (not when they're missing because the file ends in "/" though...).